File: java/src/main/java/ddp/engine/DDPEngine.java
package ddp.engine;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.io.IOException;
import java.nio.file.*;
import java.util.*;

/**
 * Thin wrapper around the ddp binary.
 * API key validation and parsing run in native code.
 */
public final class DDPEngine {
    private static final String DEFAULT_BIN = "ddp";
    private static final ObjectMapper JSON = new ObjectMapper();

    private DDPEngine() {
    }

    private static String findDDP() {
        String env = System.getenv("DDP_SDK_BIN");
        return env != null ? env : DEFAULT_BIN;
    }

    private static void runDDP(String inputPath, String outputPath, String apiKey) throws IOException {
        ProcessBuilder pb = new ProcessBuilder(
                findDDP(),
                "process",
                inputPath,
                outputPath,
                "--api-key",
                apiKey
        );
        pb.redirectErrorStream(true); // merge stderr into stdout to avoid deadlock
        Process proc = pb.start();

        String output = new String(proc.getInputStream().readAllBytes());
        try {
            int code = proc.waitFor();
            if (code != 0) {
                throw new EngineError(output, output);
            }
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            throw new EngineError("Interrupted: " + e.getMessage());
        }
    }

    private static String runDDPJson(String inputPath, String apiKey, String outputPath) throws IOException {
        List<String> args = new ArrayList<>();
        args.add(findDDP());
        args.add("json");
        args.add(inputPath);
        args.add("--api-key");
        args.add(apiKey);
        if (outputPath != null) {
            args.add("--output");
            args.add(outputPath);
        }

        ProcessBuilder pb = new ProcessBuilder(args);
        pb.redirectErrorStream(true);
        Process proc = pb.start();
        String output = new String(proc.getInputStream().readAllBytes());
        try {
            int code = proc.waitFor();
            if (code != 0) {
                throw new EngineError(output, output);
            }
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            throw new EngineError("Interrupted: " + e.getMessage());
        }
        if (outputPath != null) {
            return Files.readString(Path.of(outputPath));
        }
        return output;
    }

    /**
     * Process DDP from in-memory files. Writes metadata and WAVs to outputPath.
     * API key validation runs in the native binary. Returns metadata object.
     */
    public static JsonNode processFromBytes(Map<String, byte[]> files, String outputPath, String apiKey) throws IOException {
        Path inDir = Files.createTempDirectory("ddp-in-");
        try {
            for (Map.Entry<String, byte[]> e : files.entrySet()) {
                String filename = "SD".equals(e.getKey()) ? "SD.SD" : e.getKey();
                Files.write(inDir.resolve(filename), e.getValue());
            }
            String outBase = outputPath.replaceAll("[/\\\\]+$", "");
            runDDP(inDir.toString(), outBase, apiKey);

            String metaJson = Files.readString(Path.of(outBase, "metadata.json"));
            return JSON.readTree(metaJson);
        } finally {
            deleteRecursive(inDir);
        }
    }

    /**
     * Process DDP from a path (directory or ZIP). Invokes ddp binary.
     * API key validation runs in the native binary.
     * Returns the metadata object (metadata.json contents).
     */
    public static JsonNode process(String inputPath, String outputPath, String apiKey) throws IOException {
        String outBase = outputPath.replaceAll("[/\\\\]+$", "");
        runDDP(inputPath, outBase, apiKey);
        String metaJson = Files.readString(Path.of(outBase, "metadata.json"));
        return JSON.readTree(metaJson);
    }

    /**
     * Extract metadata JSON only (no WAV files). Invokes ddp binary.
     * API key validation runs in the native binary.
     * Returns the metadata object. If outputPath is non-null, also writes metadata.json there.
     */
    public static JsonNode processToJson(String inputPath, String apiKey, String outputPath) throws IOException {
        String jsonStr = runDDPJson(inputPath, apiKey, outputPath);
        return JSON.readTree(jsonStr);
    }

    /** Same as processToJson(inputPath, apiKey, null) — metadata only, no file write. */
    public static JsonNode processToJson(String inputPath, String apiKey) throws IOException {
        return processToJson(inputPath, apiKey, null);
    }

    private static void deleteRecursive(Path p) {
        try {
            if (Files.exists(p)) {
                Files.walk(p)
                        .sorted(Comparator.reverseOrder())
                        .forEach(path -> {
                            try {
                                Files.delete(path);
                            } catch (IOException ignored) {}
                        });
            }
        } catch (IOException ignored) {}
    }

}

File: csharp/DDPEngine/DDPEngine.cs
using System.Diagnostics;
using System.Text.Json.Nodes;

namespace DDPEngine;

/// <summary>Thin wrapper around the ddp binary. API key validation and parsing run in native code.</summary>
public static class DDPEngine
{
    private const string DefaultBin = "ddp";

    private static string FindDDP() =>
        Environment.GetEnvironmentVariable("DDP_SDK_BIN") ?? DefaultBin;

    private static void RunDDP(string inputPath, string outputPath, string apiKey)
    {
        var bin = FindDDP();
        var startInfo = new ProcessStartInfo
        {
            FileName = bin,
            ArgumentList = { "process", inputPath, outputPath, "--api-key", apiKey },
            RedirectStandardOutput = true,
            RedirectStandardError = true,
            UseShellExecute = false,
        };

        using var proc = System.Diagnostics.Process.Start(startInfo)
            ?? throw new EngineError($"Failed to start {bin}");
        proc.WaitForExit();

        var stderr = proc.StandardError.ReadToEnd();

        if (proc.ExitCode != 0)
        {
            var stdout = proc.StandardOutput.ReadToEnd();
            throw new EngineError(
                string.IsNullOrEmpty(stderr) ? stdout : stderr,
                stderr);
        }
    }

    private static string RunDDPJson(string inputPath, string apiKey, string? outputPath)
    {
        var bin = FindDDP();
        var args = new List<string> { "json", inputPath, "--api-key", apiKey };
        if (outputPath != null)
            args.AddRange(new[] { "--output", outputPath });

        var startInfo = new ProcessStartInfo
        {
            FileName = bin,
            RedirectStandardOutput = true,
            RedirectStandardError = true,
            UseShellExecute = false,
        };
        foreach (var a in args)
            startInfo.ArgumentList.Add(a);

        using var proc = System.Diagnostics.Process.Start(startInfo)
            ?? throw new EngineError($"Failed to start {bin}");
        var stdout = proc.StandardOutput.ReadToEnd();
        var stderr = proc.StandardError.ReadToEnd();
        proc.WaitForExit();

        if (proc.ExitCode != 0)
            throw new EngineError(
                string.IsNullOrEmpty(stderr) ? stdout : stderr,
                stderr);

        if (outputPath != null)
            return File.ReadAllText(outputPath);
        return stdout;
    }

    /// <summary>
    /// Process DDP from in-memory files. Writes metadata and WAVs to outputPath.
    /// API key validation runs in the native binary. Returns metadata object.
    /// </summary>
    public static JsonObject ProcessFromBytes(
        IReadOnlyDictionary<string, byte[]> files,
        string outputPath,
        string apiKey)
    {
        var inDir = Path.Combine(Path.GetTempPath(), "ddp-in-" + Guid.NewGuid().ToString("N")[..8]);
        try
        {
            Directory.CreateDirectory(inDir);
            foreach (var (name, data) in files)
            {
                var filename = name == "SD" ? "SD.SD" : name;
                File.WriteAllBytes(Path.Combine(inDir, filename), data);
            }

            var outBase = outputPath.TrimEnd('/', '\\');
            RunDDP(inDir, outBase, apiKey);

            var metaPath = Path.Combine(outBase, "metadata.json");
            var metaJson = File.ReadAllText(metaPath);
            return JsonNode.Parse(metaJson)!.AsObject();
        }
        finally
        {
            if (Directory.Exists(inDir))
                Directory.Delete(inDir, recursive: true);
        }
    }

    /// <summary>
    /// Process DDP from a path (directory or ZIP). Invokes ddp binary.
    /// API key validation runs in the native binary.
    /// Returns the metadata object (metadata.json contents).
    /// </summary>
    public static JsonObject Process(string inputPath, string outputPath, string apiKey)
    {
        var outBase = outputPath.TrimEnd('/', '\\');
        RunDDP(inputPath, outBase, apiKey);

        var metaPath = Path.Combine(outBase, "metadata.json");
        var metaJson = File.ReadAllText(metaPath);
        return JsonNode.Parse(metaJson)!.AsObject();
    }

    /// <summary>
    /// Extract metadata JSON only (no WAV files). Invokes ddp binary.
    /// API key validation runs in the native binary.
    /// Returns the metadata object. If outputPath is given, also writes metadata.json there.
    /// </summary>
    public static JsonObject ProcessToJson(string inputPath, string apiKey, string? outputPath = null)
    {
        var jsonStr = RunDDPJson(inputPath, apiKey, outputPath);
        return JsonNode.Parse(jsonStr)!.AsObject();
    }
}
